File: Array/Valid Sudoku.py
class Solution(object):
    
    
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        
        row_dic = {i:[] for i in range(9)}
        col_dic = {i:[] for i in range(9)}
        cel_dic = {i:[] for i in range(9)}
        
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    if board[i][j] in row_dic[i] \
                       or board[i][j] in col_dic[j] \
                       or board[i][j] in cel_dic[i//3 * 3 + j//3]:
                        return False
                    else:
                        row_dic[i].append(board[i][j])
                        col_dic[j].append(board[i][j])
                        cel_dic[i//3 * 3 + j//3].append(board[i][j])
                        
        return True
                        
        

File: Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<vector<int>> rowMap (9, vector<int>(9, 0));
        vector<vector<int>> colMap (9, vector<int>(9, 0));
        vector<vector<vector<int>>> cellMap(3, vector<vector<int>>(3, vector<int>(9, 0)));
        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++)
                if (board[i][j] != '.')
                {
                    int cur = board[i][j] - '1';
                    if (rowMap[i][cur] || colMap[j][cur] || cellMap[i/3][j/3][cur]) return false;
                    rowMap[i][cur] = 1;
                    colMap[j][cur] = 1;
                    cellMap[i/3][j/3][cur] = 1;
                }
        return true;
    }
};
